### PR TITLE
Add index writing capabilities

### PR DESCRIPTION
**Pull request by cryptix**

### CODE REVIEW

File: sbot/metafeed_test.go [hunk 1]
```diff
@@ -264,3 +262,170 @@ func TestMultiFeedSync(t *testing.T) {
...
+
+	// create a main feed (holds actual messages, regular old ssb feed thinger) on the root metafeed
+	mainFeedRef, err := bot.MetaFeeds.CreateSubFeed(bot.KeyPair.ID(), "main", refs.RefAlgoFeedSSB1)
+	r.NoError(err)
+
+	// note (cblgh): hmm might be nice with a <sbot>.MetaFeeds.Get(subfeedRef) (subfeed, error) method?
+
+	// create two new index feeds (distinct from the test two indexes: idx1 idx2)
+	// note (cblgh): might we introduce a method <sbot>.MetaFeeds.CreateIndex? 
+	// => creates a subfeed that is an index, with the correct `query` parameter & querylang: ssb-ql-0 set
```
Comment (cblgh) [inline sbot/metafeed_test.go:364]: this question might be answered by https://github.com/ssb-ngi-pointer/go-metafeed/pull/27 :~

**Comment by cblgh** [comment]
current status:
* this was the first place i dipped my toes into go-ssb, so there's not a lot of working code in the pr atm
* there is no `sbot.IndexFeeds` service implementation yet 
    * follow-on implication: `bot.IndexFeeds.Register` function is not implemented yet 
 
i was  fleshing out the following outline with actual statements and function calls, to the best of my go-ssb knowledge at the time :)

```golang
// * have a normal/base feed,
// * create indexes,
// * publish on the base,
// * check the index messages exist..
```

**Comment by cblgh** [comment]
@cryptix i kinda waved my arms around extensively in this [commit](https://github.com/cryptoscope/ssb/pull/114/commits/0b3f47e5a89c9d345aeff785a831dec4d133e23f ) to get the right version of go-metafeed. would be cool if you could explain (perhaps tomorrow, tuesday) yr flow for updating modules during a check-in? :) 

one thing in particular: crypto was updated, but i'm not really sure why that happened. so an explanation on that would be great

File: sbot/metafeed.go [hunk 2]
```diff
@@ -114,9 +121,43 @@ func (s metaFeedsService) CreateSubFeed(mount refs.FeedRef, purpose string, form
...
+	// we had some metadata :^)
+	if len(optionalMetadata) > 0 {
+		if len(optionalMetadata) > 1 {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: metadata to contain a single map, not multiple")
+		}
+		metadata := optionalMetadata[0]
+		// TODO: support query formats other than ssb-ql-0
+		// 2021-09-13 rationale for only supporting ssb-ql-0: we're hacking the metadata map by setting its key `query` to just contain the
+		// type, and then fixing it up on the lines below. this hack would likely fail miserably if support was added for
+		// other query languages :)
```
Comment (cryptix) [inline sbot/metafeed.go:135]: This is fine. We can't support stuff we don't know about yet.

File: sbot/metafeed.go [hunk 3]
```diff
@@ -114,9 +121,43 @@ func (s metaFeedsService) CreateSubFeed(mount refs.FeedRef, purpose string, form
...
-	return refs.FeedRef{}, fmt.Errorf("TODO: add metadata")
 	addContent := metamngmt.NewAddDerivedMessage(mountKeyPair.Feed, newSubfeedKeyPair.Feed, purpose, nonce)
 
+	// we had some metadata :^)
+	if len(optionalMetadata) > 0 {
+		if len(optionalMetadata) > 1 {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: metadata to contain a single map, not multiple")
+		}
+		metadata := optionalMetadata[0]
+		// TODO: support query formats other than ssb-ql-0
```
Comment (cryptix) [inline sbot/metafeed.go:132]:
```suggestion
```

Without defined/speced queries this is not actionable.

**Review by cryptix — COMMENTED** [review]
Looking good so far :)

**Comment by cryptix** [comment]
re https://github.com/cryptoscope/ssb/pull/114#issuecomment-918268555: yea, there was no tagged version pushed yet (`git tag vX.Y.Z && git push --tags`) for the latest metafeed. (oversight on my end, really). In this case you can do `go get $import@latest` or `go get $import@master`.

You haven't told me what you did so it's hard to guess and give better feedback but I guess you did `go get -u`? that might have updated  `x/crypto`.

**Comment by cblgh** [comment]
@cryptix ran into another strange failure state relating to the keystore in 3f06f34. the subfeed enumeration `mf.ListSubfeeds(mount)` on the root metafeed seems to be going haywire after rebooting the sbot and i couldn't figure out why. think it's best if we pair on that tomorrow when yr back or sth :)

```
    metafeed_test.go:492: 
                Error Trace:    metafeed_test.go:492
                Error:          Received unexpected error:
                                metafeed list: failed to get listing: key store: failed to get data from index: error in badger transaction (view): error getting value: error unmarshaling using json marshaler: ssb: Invalid Ref Algo
                Test:           TestMetafeedIndexes
                Messages:       failed to get index feed of restarted bot
--- FAIL: TestMetafeedIndexes (0.09s)

File: metafeeds.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+type IndexFeedManager interface {
+	// Register registers an index feed for messages of msgType on a corresponding feed contentFeed.
+	// contentFeed the feed where messages are read from
+	// output is the index feed where the index messages are published to
+	Register(indexFeed, contentFeed refs.FeedRef, msgType string) error
+	Deregister(refs.FeedRef) (bool, error)
+
+	Process(refs.Message) (refs.FeedRef, interface{}, error)
+
+	// TODO List() []refs.FeedRef
```
Comment (cryptix) [inline metafeeds.go:58]: Could you flesh out this one?

File: sbot/indexfeeds.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	"path/filepath"
+
+	"go.cryptoscope.co/ssb"
+	refs "go.mindeco.de/ssb-refs"
+)
+
+func newIndexFeedManager(storagePath string) (ssb.IndexFeedManager, error) {
+	m := indexFeedManager{indexes: make(map[string]refs.FeedRef), storagePath: storagePath}
+	// load previous registered indexes (if any)
+	m.load()
```
Comment (cryptix) [inline sbot/indexfeeds.go:22]: swallowed error.

File: sbot/indexfeeds.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+type indexFeedManager struct {
+	// registered indexes
+	indexes     map[string]refs.FeedRef
+	storagePath string
+}
+
+func constructIndexKey(author refs.FeedRef, msgType string) string {
+	return fmt.Sprintf("%s%s", author, msgType)
```
Comment (cryptix) [inline sbot/indexfeeds.go:34]:
```suggestion
	return  author.String() + msgType
```

just a stylisitic thing. also more as an generic advice: fmt format calls are pretty expensive.

File: sbot/indexfeeds.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	// creates index messages after actual messages have been published
+	var typed struct {
+		Type string
+	}
+	err := json.Unmarshal(m.ContentBytes(), &typed)
+	if err != nil {
+		return refs.FeedRef{}, nil, err
+	}
+
+	if typed.Type == "" {
```
Comment (cryptix) [inline sbot/indexfeeds.go:133]:
```suggestion
	// do nothing (reasons for no type is usually encrypted messages)
	if typed.Type == "" {
```

File: sbot/indexfeeds.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// Process looks through all the registered indexes and publishes messages accordingly
+func (manager indexFeedManager) Process(m refs.Message) (refs.FeedRef, interface{}, error) {
```
Comment (cryptix) [inline sbot/indexfeeds.go:123]:
```suggestion
func (manager indexFeedManager) Process(m refs.Message) (refs.FeedRef, IndexedMessage, error) {
```

and also return the type below ofc

File: sbot/indexfeeds.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+// Strategy: persist (as file on disk) which indexes have been created, storing contentFeed+indexFeed+type;
+// * contentFeed: the author's pubkey,
+// * indexFeed: index feed you publish to, as identified by its SSB1 feedref
+// * type: e.g. contact, about
+//
+// Currently, this is stored all in one file
+func (manager indexFeedManager) store() error {
+	data, err := json.MarshalIndent(manager.indexes, "", "  ")
+	if err != nil {
+		return fmt.Errorf("indexFeedManager marshal failed (%w)", err)
```
Comment (cryptix) [inline sbot/indexfeeds.go:64]:
```suggestion
		return fmt.Errorf("indexFeedManager marshaling indexes state failed (%w)", err)
```

File: sbot/metafeed.go [hunk 10]
```diff
@@ -32,94 +37,153 @@ func WithMetaFeedMode(enable bool) Option {
...
-	// which specifies the subfeed on which the content should be published.
-	Publish(as refs.FeedRef, content interface{}) (refs.MessageRef, error)
+		keys: keyStore,
+	}, nil
 }
 
-type SubfeedListEntry struct {
-	Feed    refs.FeedRef
-	Purpose string
+type metadataQuery struct {
```
Comment (cryptix) [inline sbot/metafeed.go:68]:
```suggestion
type MetadataQuery struct {
```

Others might want to get to the query data, so let's make this easier.

File: sbot/metafeed.go [hunk 11]
```diff
@@ -32,94 +37,153 @@ func WithMetaFeedMode(enable bool) Option {
...
+		querylang, _ := msg.GetMetadata("querylang")
+		query, exists := msg.GetMetadata("query")
+		if !exists || querylang != "ssb-ql-0" || query == "" {
+			continue
+		}
 
-	rootKeyPair metakeys.KeyPair
-}
+		// unpack the info from a string of json into something we can use
+		var queryInfo struct {
```
Comment (cryptix) [inline sbot/metafeed.go:143]: use `MetadataQuery` here?

File: sbot/metafeed.go [hunk 12]
```diff
@@ -128,59 +192,95 @@ func (s metaFeedsService) CreateSubFeed(purpose string, format refs.RefAlgo) (re
...
+
+		if _, exists := metadata["type"]; !exists {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: expected metadata to have key `type`")
+		}
+		if _, exists := metadata["author"]; !exists {
+			return refs.FeedRef{}, fmt.Errorf("CreateSubFeed: expected metadata to have key `author`")
+		}
+
+		// create a `query` object with the expected ssb-ql-0 format
+		queryJson, err := json.Marshal(metadataQuery{Author: metadata["author"], Type: metadata["type"]})
```
Comment (cryptix) [inline sbot/metafeed.go:230]: add `private: false`

**Review by cryptix — COMMENTED** [review]
Looking very good some nits and improvements.

File: metafeeds.go (hunk 4 above)
Comment (cblgh) [reply metafeeds.go:58]: sure ! do you reckon the list contents need to be sorted / stable in order?

Comment (cryptix) [reply metafeeds.go:58]: no, you can ignore the ordering but it should return the type as well.

maybe just return `[]MetaQuery`?

File: sbot/metafeed.go (hunk 12 above)
Comment (cblgh) [reply sbot/metafeed.go:230]: haha yeah, that's what i was thinking as well & why i left it out :)

File: sbot/metafeed.go (hunk 11 above)
Comment (cblgh) [reply sbot/metafeed.go:143]: good catch!

File: sbot/metafeed_test.go [hunk 13]
```diff
@@ -485,6 +490,18 @@ func TestMetafeedIndexes(t *testing.T) {
...
+	r.NoError(err)
+	r.Len(indexlist, 1, "expected number of registered indexes to be 1")
+
+	/* perform some index listing operations */
+	indexlist, err = bot.IndexFeeds.ListByType("vote")
+	r.NoError(err)
+	r.Len(indexlist, 0, "expected no indexes of type vote")
+	indexlist, err = bot.IndexFeeds.ListByType("contact")
+	r.NoError(err)
+	r.Len(indexlist, 1, "expected number of registered indexes of type contact to be 1")
```
Comment (cryptix) [inline sbot/metafeed_test.go:504]: <3

File: sbot/indexfeeds.go [hunk 14]
```diff
@@ -120,44 +127,63 @@ func (manager indexFeedManager) Deregister(indexFeed refs.FeedRef) (bool, error)
...
-		Indexed: indexed{
-			m.Seq(),
-			m.Key(),
-		},
-	}
+	content := ssb.IndexedMessage{ Type: "indexed" }
+	content.Indexed.Sequence = m.Seq()
+	content.Indexed.Key=  m.Key()
 
+	manager.List() // TODO remove this cause it's just for testing
```
Comment (cryptix) [inline sbot/indexfeeds.go]: 😬

File: sbot/indexfeeds.go [hunk 15]
```diff
@@ -120,44 +127,63 @@ func (manager indexFeedManager) Deregister(indexFeed refs.FeedRef) (bool, error)
...
-	Key      refs.MessageRef `json:"key"`
+func (manager indexFeedManager) List() ([]ssb.IndexListEntry, error) {
+	return manager.ListByType("")
 }
-type indexMsg struct {
-	Type    string  `json:"type"`
-	Indexed indexed `json:"indexed"`
+
+func (manager indexFeedManager) ListByType(msgType string) ([]ssb.IndexListEntry, error) {
+	indexlist := make([]ssb.IndexListEntry, 0, len(manager.indexes))
```
Comment (cryptix) [inline sbot/indexfeeds.go:164]: if you want to be space/allocation efficient here, you should rather allocate the  whole thing, keep and index when filling and set the fields directly. Sadly the compiler isn't smart enough to figure out the way you create values below and append them can be done without copying them over.

```go
foo := make([]mystruct, 23}
fooIdx := 0

for ... := range {
  // filter etc
  foo[fooIdx].meta = m
  foo[fooIdx].stuff = s
  fooIdx++
}
```

**Review by cryptix — COMMENTED** [review]
some nitpicks but nothing major! 

think we can merge after this :)

File: sbot/indexfeeds.go (hunk 14 above)
Comment (cblgh) [reply sbot/indexfeeds.go]: already gone ;)

https://github.com/cryptoscope/ssb/pull/114/commits/d26191dfde8c974b41646cfde9b5564212935f31

File: sbot/indexfeeds.go (hunk 15 above)
Comment (cblgh) [reply sbot/indexfeeds.go:164]: hmm that's a shame! thanks for letting me know c: i'll implement your suggested tweak (tho in this particular case it seems unlikely to matter tooo much from a performance pov?)

Comment (cryptix) [reply sbot/indexfeeds.go:164]: you are right. this isn't perf ciritical but i've seen you doing this before.

as a matter of fact, in this case you could also leave out the `make([]stuff, 0, cap)` and just declare it as `var s []stuff` and use append like you did before and have it grow as needed with each loop.

Comment (cblgh) [reply sbot/indexfeeds.go]: you are 100% correct i plead guilty to being attached to `append(list, item)` 😳

i appreciate you questioning that pattern, i think i'm walking away with a slightly better way of thinking about this type of situation now :3

**Comment by cblgh** [comment]
@cryptix squashed & redo to go, i think! maybe glance over https://github.com/cryptoscope/ssb/pull/114/commits/a71fc8aeb96eeb002bb4960a2585a82bf10dd8b6 to make sure it's alright :)

**Comment by cblgh** [comment]
woopa woop!

**Pull request merged**